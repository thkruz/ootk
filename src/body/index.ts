export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
