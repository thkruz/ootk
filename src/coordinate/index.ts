export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
