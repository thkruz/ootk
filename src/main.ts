/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
