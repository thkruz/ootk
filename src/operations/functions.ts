/* eslint-disable require-jsdoc */
/* eslint-disable func-style */
import { Matrix } from './Matrix';
import { Vector } from './Vector';

export type DifferentiableFunction = (x: number) => number;
export type JacobianFunction = (xs: Float64Array) => Float64Array;

export enum AngularDistanceMethod {
  Cosine,
  Haversine,
}

export enum AngularDiameterMethod {
  Circle,
  Sphere,
}

export function factorial(n: number): number {
  const nAbs = Math.abs(n);
  let result = 1;

  for (let i = 2; i <= nAbs; i++) {
    result *= i;
  }

  return result;
}

export function log10(x: number): number {
  return Math.log(x) / Math.LN10;
}

export function cbrt(n: number): number {
  return n ** (1 / 3);
}

export function sinh(x: number): number {
  return 0.5 * (Math.exp(x) - Math.exp(-x));
}

export function cosh(x: number): number {
  return 0.5 * (Math.exp(x) + Math.exp(-x));
}

export function tanh(x: number): number {
  return sinh(x) / cosh(x);
}

export function coth(x: number): number {
  return cosh(x) / sinh(x);
}

export function sech(x: number): number {
  return 1 / cosh(x);
}

export function csch(x: number): number {
  return 1 / sinh(x);
}

export function asinh(x: number): number {
  return Math.log(x + Math.sqrt(x * x + 1));
}

export function acosh(x: number): number {
  return Math.log(x + Math.sqrt(x * x - 1));
}

export function atanh(x: number): number {
  return 0.5 * Math.log((1 + x) / (1 - x));
}

export function acsch(x: number): number {
  return Math.log(1 / x + Math.sqrt(1 / (x * x) + 1));
}

export function asech(x: number): number {
  return Math.log(1 / x + Math.sqrt(1 / (x * x) - 1));
}

export function acoth(x: number): number {
  return 0.5 * Math.log((x + 1) / (x - 1));
}

export function copySign(mag: number, sgn: number): number {
  return Math.abs(mag) * Math.sign(sgn);
}

export function evalPoly(x: number, coeffs: Float64Array): number {
  let result = coeffs[0];

  for (let i = 1; i < coeffs.length; i++) {
    result = result * x + coeffs[i];
  }

  return result;
}

export function concat(a: Float64Array, b: Float64Array): Float64Array {
  const result = new Float64Array(a.length + b.length);

  result.set(a);
  result.set(b, a.length);

  return result;
}

export function matchHalfPlane(angle: number, match: number): number {
  const a1 = angle;
  const a2 = 2 * Math.PI - angle;
  const d1 = Math.atan2(Math.sin(a1 - match), Math.cos(a1 - match));
  const d2 = Math.atan2(Math.sin(a2 - match), Math.cos(a2 - match));

  return Math.abs(d1) < Math.abs(d2) ? a1 : a2;
}

export function wrapAngle(theta: number): number {
  const result = ((theta + Math.PI) % (2 * Math.PI)) - Math.PI;

  if (result === -Math.PI) {
    return Math.PI;
  }

  return result;
}

export function angularDistanceCosine(lam1: number, phi1: number, lam2: number, phi2: number): number {
  const a = Math.sin(phi1) * Math.sin(phi2);
  const b = Math.cos(phi1) * Math.cos(phi2) * Math.cos(lam2 - lam1);

  return Math.acos(a + b);
}

export function angularDistanceHaversine(lam1: number, phi1: number, lam2: number, phi2: number): number {
  const dlam = lam2 - lam1;
  const dphi = phi2 - phi1;
  const sdlam = Math.sin(0.5 * dlam);
  const sdphi = Math.sin(0.5 * dphi);
  const a = sdphi * sdphi + Math.cos(phi1) * Math.cos(phi2) * sdlam * sdlam;

  return 2.0 * Math.asin(Math.min(1.0, Math.sqrt(a)));
}

export function angularDistance(
  lam1: number,
  phi1: number,
  lam2: number,
  phi2: number,
  method: AngularDistanceMethod = AngularDistanceMethod.Cosine,
): number {
  switch (method) {
    case AngularDistanceMethod.Cosine:
      return angularDistanceCosine(lam1, phi1, lam2, phi2);
    case AngularDistanceMethod.Haversine:
      return angularDistanceHaversine(lam1, phi1, lam2, phi2);
    default:
      throw new Error('Invalid angular distance method.');
  }
}

export function angularDiameter(
  diameter: number,
  distance: number,
  method: AngularDiameterMethod = AngularDiameterMethod.Sphere,
): number {
  switch (method) {
    case AngularDiameterMethod.Circle:
      return 2 * Math.atan(diameter / (2 * distance));
    case AngularDiameterMethod.Sphere:
      return 2 * Math.asin(diameter / (2 * distance));
    default:
      throw new Error('Invalid angular diameter method.');
  }
}

export function linearInterpolate(x: number, x0: number, y0: number, x1: number, y1: number): number {
  return (y0 * (x1 - x) + y1 * (x - x0)) / (x1 - x0);
}

export function mean(values: number[]): number {
  const n = values.length;
  let sum = 0.0;

  for (const v of values) {
    sum += v;
  }

  return sum / n;
}

export function standardDeviation(values: number[], isSample = false): number {
  const mu = mean(values);
  const n = values.length;
  let sum = 0.0;

  for (const v of values) {
    const sub = v - mu;

    sum += sub * sub;
  }
  const m = isSample ? 1 : 0;

  return Math.sqrt((1.0 / (n - m)) * sum);
}

export function covariance(a: number[], b: number[], isSample = false): number {
  const n = a.length;
  const am = mean(a);
  const bm = mean(b);
  let result = 0.0;

  for (let i = 0; i < n; i++) {
    result += (a[i] - am) * (b[i] - bm);
  }
  const m = isSample ? 1 : 0;

  return result / (n - m);
}

export function derivative(f: DifferentiableFunction, h = 1e-3): DifferentiableFunction {
  function df(x: number): number {
    const hh = h * 0.5;

    return (f(x + hh) - f(x - hh)) / h;
  }

  return df;
}

export function gamma(n: number): number {
  return factorial(n - 1);
}

export function newtonM(ecc: number, m: number): { e0: number; nu: number } {
  const numiter = 50;
  const small = 1e-8;
  let e0;
  let nu;

  if (ecc > small) {
    if ((m < 0.0 && m > -Math.PI) || m > Math.PI) {
      e0 = m - ecc;
    } else {
      e0 = m + ecc;
    }
    let ktr = 1;
    let e1 = e0 + (m - e0 + ecc * Math.sin(e0)) / (1.0 - ecc * Math.cos(e0));

    while (Math.abs(e1 - e0) > small && ktr <= numiter) {
      ktr++;
      e0 = e1;
      e1 = e0 + (m - e0 + ecc * Math.sin(e0)) / (1.0 - ecc * Math.cos(e0));
    }
    const sinv = (Math.sqrt(1.0 - ecc * ecc) * Math.sin(e1)) / (1.0 - ecc * Math.cos(e1));
    const cosv = (Math.cos(e1) - ecc) / (1.0 - ecc * Math.cos(e1));

    nu = Math.atan2(sinv, cosv);
  } else {
    nu = m;
    e0 = m;
  }

  return { e0, nu };
}

export function newtonNu(ecc: number, nu: number): { e0: number; m: number } {
  const small = 1e-8;
  let e0 = 0.0;
  let m = 0.0;

  if (Math.abs(ecc) < small) {
    m = nu;
    e0 = nu;
  } else if (ecc < 1.0 - small) {
    const sine = (Math.sqrt(1.0 - ecc * ecc) * Math.sin(nu)) / (1.0 + ecc * Math.cos(nu));
    const cose = (ecc + Math.cos(nu)) / (1.0 + ecc * Math.cos(nu));

    e0 = Math.atan2(sine, cose);
    m = e0 - ecc * Math.sin(e0);
  }
  if (ecc < 1.0) {
    m -= Math.floor(m / (2 * Math.PI)) * (2 * Math.PI);
    if (m < 0.0) {
      m += 2.0 * Math.PI;
    }
    e0 -= Math.floor(e0 / (2 * Math.PI)) * (2 * Math.PI);
  }

  return { e0, m };
}

export function array2d<T>(rows: number, columns: number, value: T): T[][] {
  const output: T[][] = [];

  for (let i = 0; i < rows; i++) {
    output.push(Array(columns).fill(value));
  }

  return output;
}

export function jacobian(f: JacobianFunction, m: number, x0: Float64Array, step = 1e-5): Matrix {
  const n = x0.length;
  const j = array2d(m, n, 0.0);
  const h = 0.5 * step;

  for (let k = 0; k < n; k++) {
    const xp = x0.slice();

    xp[k] += h;
    const fp = new Vector(f(xp));

    const xm = x0.slice();

    xm[k] -= h;
    const fm = new Vector(f(xm));

    const cd = fp.subtract(fm).scale(1.0 / step);

    for (let i = 0; i < m; i++) {
      j[i][k] = cd[i];
    }
  }

  return new Matrix(j);
}

/**
 * Clamps a number between a minimum and maximum value.
 * @param x The number to clamp.
 * @param min The minimum value.
 * @param max The maximum value.
 * @returns The clamped number.
 */
export function clamp(x: number, min: number, max: number): number {
  return Math.max(min, Math.min(x, max));
}
