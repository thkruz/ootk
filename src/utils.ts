/**
 * @author Theodore Kruczek.
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects.
 *
 * @file The Utils module.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

type vec3K = {
  x: number;
  y: number;
  z: number;
};

class Utils {
  public static distance(pos1: vec3K, pos2: vec3K): number {
    return Math.sqrt((pos1.x - pos2.x) ** 2 + (pos1.y - pos2.y) ** 2 + (pos1.z - pos2.z) ** 2);
  }

  public static dopplerFactor(location: vec3K, position: vec3K, velocity: vec3K): number {
    const mfactor = 7.292115e-5;
    const c = 299792.458; // Speed of light in km/s

    const range = {
      x: position.x - location.x,
      y: position.y - location.y,
      z: position.z - location.z,
      w: 0,
    };
    range.w = Math.sqrt(range.x ** 2 + range.y ** 2 + range.z ** 2);

    const rangeVel = {
      x: velocity.x + mfactor * location.y,
      y: velocity.y - mfactor * location.x,
      z: velocity.z,
    };

    const sign = (value) => (value >= 0 ? 1 : -1);

    const rangeRate =
      (range.x * rangeVel.x + range.y * rangeVel.y + range.z * rangeVel.z) / range.w;

    return 1 + (rangeRate / c) * sign(rangeRate);
  }

  public static createVec(start: number, stop:number, step:number): number[] {
    let array = [];
    for (let i = start; i <= stop; i += step) {
      array.push(i);
    }
    return array;
  }
}

export { Utils };
