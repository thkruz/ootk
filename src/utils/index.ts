export * from './jacobian.js';
export * from './functions.js';
export * from './constants.js';
export * from './linearDistance.js';
