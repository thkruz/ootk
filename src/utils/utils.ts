/**
 * @author Theodore Kruczek.
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects.
 *
 * @file The Utils module.
 *
 * @license AGPL-3.0-or-later
 * @Copyright (c) 2020-2022 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import * as Types from '../types/types';
import { DAY_TO_MS } from './constants';
import { MoonMath } from './moon-math';
import { SunMath } from './sun-math';

class Utils {
  public static Types = Types;

  public static distance(pos1: Types.EciVec3, pos2: Types.EciVec3): Types.Kilometer {
    return Math.sqrt((pos1.x - pos2.x) ** 2 + (pos1.y - pos2.y) ** 2 + (pos1.z - pos2.z) ** 2);
  }

  private static sign = (value: number) => (value >= 0 ? 1 : -1);

  public static dopplerFactor(
    location: Types.EciVec3,
    position: Types.EciVec3,
    velocity: Types.EciVec3,
  ): Types.Kilometer {
    const mfactor = 7.292115e-5;
    const c = 299792.458; // Speed of light in km/s

    const range = <Types.EciVec3>{
      x: position.x - location.x,
      y: position.y - location.y,
      z: position.z - location.z,
    };
    const distance = Math.sqrt(range.x ** 2 + range.y ** 2 + range.z ** 2);

    const rangeVel = <Types.EciVec3>{
      x: velocity.x + mfactor * location.y,
      y: velocity.y - mfactor * location.x,
      z: velocity.z,
    };

    const rangeRate = (range.x * rangeVel.x + range.y * rangeVel.y + range.z * rangeVel.z) / distance;

    return 1 + (rangeRate / c) * Utils.sign(rangeRate);
  }

  public static createVec(start: number, stop: number, step: number): number[] {
    const array = [];

    for (let i = start; i <= stop; i += step) {
      array.push(i);
    }

    return array;
  }

  // eslint-disable-next-line max-params
  public static jday = (year?: number, mon?: number, day?: number, hr?: number, minute?: number, sec?: number) => {
    if (!year) {
      const now = new Date();
      const jDayStart = new Date(now.getUTCFullYear(), 0, 0);
      const jDayDiff = now.getDate() - jDayStart.getDate();

      return Math.floor(jDayDiff / DAY_TO_MS);
    }

    return (
      367.0 * year -
      Math.floor(7 * (year + Math.floor((mon + 9) / 12.0)) * 0.25) +
      Math.floor((275 * mon) / 9.0) +
      day +
      1721013.5 +
      ((sec / 60.0 + minute) / 60.0 + hr) / 24.0
    );
  };

  public static roundToNDecimalPlaces(value: number, places: number): number {
    return Math.round(value * 10 ** places) / 10 ** places;
  }

  public static SunMath = SunMath;
  public static MoonMath = MoonMath;
}

export { Utils };
