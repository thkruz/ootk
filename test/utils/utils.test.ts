import { EcfVec3, Kilometers, KilometersPerSecond, linearDistance, RADIUS_OF_EARTH, Vec3 } from '../../src/main';
import { dopplerFactor, getDayOfYear } from '../../src/utils/functions';

const sincos45deg = Math.sqrt(2) / 2;

describe('Doppler factor', () => {
  it('works without observer movement and object moving away', () => {
    // North Pole
    const observerEcf = {
      x: 0,
      y: 0,
      z: RADIUS_OF_EARTH,
    } as EcfVec3;
    const positionEcf = {
      x: 0,
      y: 0,
      z: RADIUS_OF_EARTH + 500,
    } as EcfVec3;
    const velocityEcf = {
      x: 0,
      y: 0,
      z: 1,
    } as EcfVec3<KilometersPerSecond>;
    const dopFactor = dopplerFactor(observerEcf, positionEcf, velocityEcf);

    expect(dopFactor).toBeLessThan(1);
  });

  it('works without observer movement and object moving towards', () => {
    // North Pole
    const observerEcf = {
      x: 0,
      y: 0,
      z: RADIUS_OF_EARTH,
    } as EcfVec3;
    const positionEcf = {
      x: 0,
      y: 0,
      z: RADIUS_OF_EARTH + 500,
    } as EcfVec3;
    const velocityEcf = {
      x: 0,
      y: 0,
      z: -1,
    } as EcfVec3<KilometersPerSecond>;
    const dopFactor = dopplerFactor(observerEcf, positionEcf, velocityEcf);

    expect(dopFactor).toBeGreaterThan(1);
  });

  it('calculates earth rotation the same as before #1', () => {
    const observerEcf = {
      x: RADIUS_OF_EARTH,
      y: 0,
      z: 0,
    } as EcfVec3;
    const positionEcf = {
      x: (RADIUS_OF_EARTH + 500) * sincos45deg, // z*sin(45)
      y: (RADIUS_OF_EARTH + 500) * sincos45deg, // z*cos(45)
      z: 0,
    } as EcfVec3;
    const velocityEcf = {
      x: 0,
      y: 0,
      z: 0,
    } as EcfVec3<KilometersPerSecond>;
    const dopFactor = dopplerFactor(observerEcf, positionEcf, velocityEcf);

    expect(dopFactor).toMatchSnapshot();
  });

  it('works without observer movement and a stationary object', () => {
    // North Pole
    const observerEcf = {
      x: 0,
      y: 0,
      z: RADIUS_OF_EARTH,
    } as EcfVec3;
    const positionEcf = {
      x: 0,
      y: 0,
      z: RADIUS_OF_EARTH + 500,
    } as EcfVec3;
    const velocityEcf = {
      x: 0,
      y: 0,
      z: 0,
    } as EcfVec3<KilometersPerSecond>;
    const dopFactor = dopplerFactor(observerEcf, positionEcf, velocityEcf);

    expect(dopFactor).toEqual(1);
  });

  it('calculates earth rotation the same as before #2', () => {
    // North Pole
    const observerEcf = {
      x: RADIUS_OF_EARTH,
      y: 0,
      z: 0,
    } as EcfVec3;
    const positionEcf = {
      x: RADIUS_OF_EARTH + 500,
      y: 0,
      z: 0,
    } as EcfVec3;
    const velocityEcf = {
      x: 0,
      y: 0,
      z: 1,
    } as EcfVec3<KilometersPerSecond>;
    const dopFactor = dopplerFactor(observerEcf, positionEcf, velocityEcf);

    expect(dopFactor).toMatchSnapshot();
  });
});

describe('Distance function', () => {
  test('if distance calculation is correct', () => {
    expect(
      linearDistance(
        { x: 1000, y: 1000, z: 1000 } as Vec3<Kilometers>,
        { x: 1000, y: 1000, z: 1000 } as Vec3<Kilometers>,
      ),
    ).toEqual(0);
    expect(
      linearDistance(
        { x: 1000, y: 1000, z: 1000 } as Vec3<Kilometers>,
        { x: 1000, y: 1000, z: 1100 } as Vec3<Kilometers>,
      ),
    ).toEqual(100);
  });
});

describe('doy Functions', () => {
  test('if doy is correct', () => {
    expect(getDayOfYear(new Date(2022, 0, 1))).toEqual(1);
    expect(getDayOfYear(new Date(2022, 1, 1))).toEqual(32);
    expect(getDayOfYear(new Date(2022, 2, 1))).toEqual(60);
    expect(getDayOfYear(new Date(2022, 3, 1))).toEqual(91);
    expect(getDayOfYear(new Date(2022, 4, 1))).toEqual(121);
    expect(getDayOfYear(new Date(2022, 5, 1))).toEqual(152);
    expect(getDayOfYear(new Date(2022, 6, 1))).toEqual(182);
    expect(getDayOfYear(new Date(2022, 7, 1))).toEqual(213);
    expect(getDayOfYear(new Date(2022, 8, 1))).toEqual(244);
    expect(getDayOfYear(new Date(2022, 9, 1))).toEqual(274);
    expect(getDayOfYear(new Date(2022, 10, 1))).toEqual(305);
    expect(getDayOfYear(new Date(2022, 11, 1))).toEqual(335);
  });

  test('if getDayOfYear is correct in leap year', () => {
    expect(getDayOfYear(new Date(2020, 0, 1, 0))).toEqual(1);
    expect(getDayOfYear(new Date(2020, 1, 1, 0))).toEqual(32);
    expect(getDayOfYear(new Date(2020, 2, 1, 0))).toEqual(61);
    expect(getDayOfYear(new Date(2020, 3, 1, 0))).toEqual(92);
    expect(getDayOfYear(new Date(2020, 4, 1, 0))).toEqual(122);
    expect(getDayOfYear(new Date(2020, 5, 1, 0))).toEqual(153);
    expect(getDayOfYear(new Date(2020, 6, 1, 0))).toEqual(183);
    expect(getDayOfYear(new Date(2020, 7, 1, 0))).toEqual(214);
    expect(getDayOfYear(new Date(2020, 8, 1, 0))).toEqual(245);
    expect(getDayOfYear(new Date(2020, 9, 1, 0))).toEqual(275);
    expect(getDayOfYear(new Date(2020, 10, 1, 0))).toEqual(306);
    expect(getDayOfYear(new Date(2020, 11, 1, 0))).toEqual(336);
  });
});
