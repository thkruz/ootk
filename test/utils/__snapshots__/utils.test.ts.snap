// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doppler factor calculates earth rotation the same as before #1 1`] = `1.000001479636888`;

exports[`Doppler factor calculates earth rotation the same as before #2 1`] = `1`;
